datalad/cmdline/main.py

```diff
@@ -277,18 +319,63 @@ def setup_parser(
             if hasattr(_intf, 'result_renderer_cmdline'):
                 plumbing_args['result_renderer'] = _intf.result_renderer_cmdline
             subparser.set_defaults(**plumbing_args)
+            parts[cmd_name] = subparser
             # store short description for later
             sdescr = getattr(_intf, 'short_description',
                              parser_args['description'].split('\n')[0])
             cmd_short_descriptions.append((cmd_name, sdescr))
-            parts[cmd_name] = subparser
         grp_short_descriptions.append(cmd_short_descriptions)
 
     # create command summary
+    if '--help' in cmdlineargs or '--help-np' in cmdlineargs:
+        parser.description = get_description_with_cmd_summary(
+            grp_short_descriptions,
+            interface_groups,
+            parser.description)
+
+    parts['datalad'] = parser
+    lgr.log(5, "Finished setup_parser")
+    if return_subparsers:
+        return parts
+    else:
+        return parser
+
+
+def fail_with_short_help(parser,
+                         msg=None,
+                         known=None, provided=None,
+                         hint=None,
+                         exit_code=1,
+                         what="command"):
+    """Generic helper to fail
+    with short help possibly hinting on what was intended if `known`
+    were provided
+    """
+    if msg:
+        sys.stderr.write("error: %s\n" % msg)
+    if not known:
+        # just to appear in print_usage also consistent with --help output
+        parser.add_argument("command [command-opts]")
+        parser.print_usage(file=sys.stderr)
+    else:
+        sys.stderr.write("Unknown %s %s.\n" % (what, provided,))
+        import difflib
+        suggestions = difflib.get_close_matches(provided, known)
+        if suggestions:
+            sys.stderr.write(" Did you mean %s?\n" % " or ".join(suggestions))
+        sys.stderr.write(" All known %ss: %s\n" % (what, ", ".join(sorted(known))))
```

Comment:
I wonder if we should bother filling the screen up with all known commands.

Here's what git does:

```
% git dff
git: 'dff' is not a git command. See 'git --help'.

The most similar command is
        diff
```

And git-annex

```
% git annex addulr
git-annex: Unknown command 'addulr'

Did you mean one of these?
        addurl
        adjust

```

Reply:
ok, married the two (but didn't use annex'es "dump all commands" if no suggestion found):
```shell
$> datalad lkjsldkfjsdf
datalad: Unknown command 'lkjsldkfjsdf'.  See 'datalad --help'.

$> datalad installl    
datalad: Unknown command 'installl'.  See 'datalad --help'.

Did you mean one of these?
        install
        uninstall
```
and also will use "known" extensions in suggestions:
```shell
$> datalad crawll
datalad: Unknown command 'crawll'.  See 'datalad --help'.
 
Did you mean one of these?
        crawl
        crawl-init

$> datalad crawl   
datalad: Unknown command 'crawl'.  See 'datalad --help'.

Hint: Command crawl is provided by (not installed) extension datalad-crawler.
```

Will add a basic test and push...